File: 11/one.rb
# frozen_string_literal: true

require 'pathname'
require 'pry-nav'

arg = ARGV.shift

case arg
when /input|input.txt/
  FILE_PATH = Pathname.new(File.expand_path('input.txt'))
when /test|test.txt/
  FILE_PATH = Pathname.new(File.expand_path('test.txt'))
else
  raise ArgumentError, "Expects 'input' or 'test' as command line argument"
end

lines = File.open(FILE_PATH, File::RDONLY).readlines(chomp: true)
nums = lines.first.split.map(&:to_i)

def blink(nums)
  nums.map do |num|
  # If the stone is engraved with the number 0, it is replaced by a stone engraved with the number 1.
  if num == 0
    1
  # If the stone is engraved with a number that has an even number of digits, it is replaced by two stones.
  # The left half of the digits are engraved on the new left stone,
  # and the right half of the digits are engraved on the new right stone.
  # (The new numbers don't keep extra leading zeroes: 1000 would become stones 10 and 0.)
  elsif num.to_s.length.even?
    s = num.to_s
    half = s.length / 2
    left = s[0...half]
    right = s[half..]
    [left.to_i, right.to_i]

  # If none of the other rules apply, the stone is replaced by a new stone;
  # the old stone's number multiplied by 2024 is engraved on the new stone.
  else
    num * 2024
  end

  end.flatten
end

25.times do
  nums = blink(nums)
end
p nums.length

File: 11/one.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

func main() {
	args := os.Args[1:]
	if len(args) != 1 {
		panic("error: expects 'input' or 'test' as an argument")
	}

	var filepath string
	arg := args[0]
	if matched, err := regexp.Match("test", []byte(arg)); matched {
		if err != nil {
			panic(err)
		}
		filepath = "test.txt"
	} else if matched, err := regexp.Match("input", []byte(arg)); matched {
		if err != nil {
			panic(err)
		}
		filepath = "input.txt"
	} else {
		panic("error: expects 'input' or 'test' as an argument")
	}

	file, err := os.Open(filepath)
	if err != nil {
		panic(err)
	}

	scanner := bufio.NewScanner(file)
	lines := []string{}
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	nums := []int{}
	for _, numStr := range regexp.MustCompile(`\s+`).Split(lines[0], -1) {
		num, err := strconv.Atoi(numStr)
		if err != nil {
			panic(err)
		}
		nums = append(nums, num)
	}

	for i := 0; i < 25; i++ {
		newNums := make([]int, 0, len(nums))
		for j := range nums {
			children := blink(nums[j])
			for _, child := range children {
				newNums = append(newNums, child)
			}
		}
		nums = newNums
	}

	fmt.Println(len(nums))
}

func blink(num int) (res []int) {
	if num == 0 {
		res = []int{1}
	} else if numStr := strconv.Itoa(num); len(numStr)&1 == 0 {
		half := len(numStr) / 2
		leftStr := numStr[0:half]
		rightStr := numStr[half:]

		left, err := strconv.Atoi(leftStr)
		if err != nil {
			panic(err)
		}

		right, err := strconv.Atoi(rightStr)
		if err != nil {
			panic(err)
		}
		res = []int{left, right}
	} else {
		res = []int{num * 2024}
	}
	return res
}

// answer: 193269